core/rawdb/accessors_chain.go

```diff
@@ -79,6 +79,33 @@ func ReadAllHashes(db ethdb.Iteratee, number uint64) []common.Hash {
 	return hashes
 }
 
+// ReadAllCanonicalHashes retrieves all canonical number and hash mappings at the
+// certain chain range. Note, this method should only used in pruned light client
+// otherwise the cost can be very expensive.
+func ReadAllCanonicalHashes(db ethdb.Iteratee, from uint64, to uint64) ([]uint64, []common.Hash) {
+	var (
+		numbers []uint64
+		hashes  []common.Hash
+	)
+	it := db.NewIteratorWithPrefix(headerPrefix)
+	defer it.Release()
+
+	for it.Next() {
+		if key := it.Key(); len(key) == len(headerPrefix)+8+1 && bytes.Equal(key[len(key)-1:], headerHashSuffix) {
```

Comment:
You should also break the loop here if the key does not begin with `headerPrefix`. If the last header number is not larger than `to` then this will iterate through the entire db.

Reply:
I use `PrefixIterator` here. So all returned entries should have prefix `headerPrefix`(no checking needed).

I guess you want to use `StartIterator` here to improve the efficiency here. But as you said since usually we will start iteration from block1, so it won't help too much. Besides, if we want `StartIterator`, we have to check whether the key starts with `headerPrefix`, the current implementation seems easier.

Reply:
Right, my mistake :) It is fine then.